operator/builtin/input/azure/loganalytics/parse.go

```diff
@@ -0,0 +1,168 @@
+	l.WG.Add(1)
+	defer l.WG.Done()
+
+	type record struct {
+		Records []map[string]interface{} `json:"records"`
+	}
+
+	// Create a "base" event by capturing the batch log records from the event's Data field.
+	// If Unmarshalling fails, fallback on handling the event as a single log entry.
+	records := record{}
+	if err := json.Unmarshal(event.Data, &records); err != nil {
+		id := event.ID
+		if id == "" {
+			event.ID = "unknown"
+		}
+		l.Warnw(fmt.Sprintf("Failed to parse event '%s' as JSON. Expcted key 'records' in event.Data.", string(event.Data)), zap.Error(err))
+		l.handleEvent(ctx, *event, nil)
+		return nil
+	}
+	event.Data = nil
+
+	// Create an entry for each log in the batch, using the origonal event's fields
+	// as a starting point for each entry
+	wg := sync.WaitGroup{}
+	max := 10
+	guard := make(chan struct{}, max)
+	for i := 0; i < len(records.Records); i++ {
+		r := records.Records[i]
+		wg.Add(1)
+		guard <- struct{}{}
+		go func() {
+			defer func() {
+				wg.Done()
+				<-guard
+			}()
+			l.handleEvent(ctx, *event, r)
+		}()
+	}
+	wg.Wait()
+	return nil
+}
+
+func (l *LogAnalyticsInput) handleEvent(ctx context.Context, event azhub.Event, records map[string]interface{}) {
+	e, err := l.NewEntry(nil)
+	if err != nil {
+		l.Errorw("Failed to parse event as an entry", zap.Error(err))
+		return
+	}
+
+	if err = l.parse(event, records, e); err != nil {
+		l.Errorw("Failed to parse event as an entry", zap.Error(err))
+		return
+	}
+	l.Write(ctx, e)
+}
+
+// parse returns an entry from an event and set of records
+func (l *LogAnalyticsInput) parse(event azhub.Event, records map[string]interface{}, e *entry.Entry) error {
+	// make sure all keys are lower case
+	for k, v := range records {
+		delete(records, k)
+		records[strings.ToLower(k)] = v
+	}
+
+	// Add base fields shared among all log records from the event
+	err := azure.ParseEvent(event, e)
+	if err != nil {
+		return err
+	}
+
+	// set label azure_log_analytics_table
+	records, err = l.setType(e, records)
+	if err != nil {
+		return err
+	}
+
+	if err := l.setTimestamp(e, records); err != nil {
+		return err
+	}
+
+	// Add remaining records to record.<azure_log_analytics_table> map
+	return l.setField(e, e.Attributes["azure_log_analytics_table"], records)
+}
+
+// setType sets the label 'azure_log_analytics_table'
+func (l *LogAnalyticsInput) setType(e *entry.Entry, records map[string]interface{}) (map[string]interface{}, error) {
+	const typeField = "type"
+
+	for key, value := range records {
+		if key == typeField {
+			if v, ok := value.(string); ok {
+				v = strings.ToLower(v)
+
+				// Set the log table label
+				if err := l.setLabel(e, "azure_log_analytics_table", v); err != nil {
+					return nil, err
+				}
+
+				delete(records, key)
+				return records, nil
+			}
+			return nil, fmt.Errorf("expected '%s' field to be a string", typeField)
+		}
+	}
+	return nil, fmt.Errorf("expected to find field with name '%s'", typeField)
+}
```

Comment:
Exit early to flatten

```suggestion
func (l *LogAnalyticsInput) setType(e *entry.Entry, records map[string]interface{}) (map[string]interface{}, error) {
	const typeField = "type"

	for key, value := range records {
		if key != typeField {
			continue
		}
		
		v, ok := value.(string)
		if !ok {
			return nil, fmt.Errorf("expected '%s' field to be a string", typeField)
		}
		
		v = strings.ToLower(v)

		// Set the log table label
		if err := l.setLabel(e, "azure_log_analytics_table", v); err != nil {
			return nil, err
		}

		delete(records, key)
		return records, nil
	}
	return nil, fmt.Errorf("expected to find field with name '%s'", typeField)
}
```

Reply:
Much better, I commited this.